File: frontend/src/utils/connectionReadOnly.ts
import type { ConnectionConfig } from "../types";
import { convertMongoShellToJsonCommand } from "./mongodb";
import { resolveSqlDialect } from "./sqlDialect";
import { findSqlStatementRanges } from "./sqlStatementSelection";

export type ConnectionProtectionKey =
  | "restrictDataEdit"
  | "restrictStructureEdit"
  | "restrictScriptExecution"
  | "restrictDataImport";

export type ConnectionProtectionConfig = NonNullable<
  ConnectionConfig["protection"]
>;

type ConnectionReadOnlyLike = Pick<
  ConnectionConfig,
  "type" | "driver" | "oceanBaseProtocol" | "readOnly" | "protection"
> | null | undefined;

export const CONNECTION_PROTECTION_KEYS: ConnectionProtectionKey[] = [
  "restrictDataEdit",
  "restrictStructureEdit",
  "restrictScriptExecution",
  "restrictDataImport",
];

const EMPTY_CONNECTION_PROTECTION: ConnectionProtectionConfig = {
  restrictDataEdit: false,
  restrictStructureEdit: false,
  restrictScriptExecution: false,
  restrictDataImport: false,
};

const FULL_CONNECTION_PROTECTION: ConnectionProtectionConfig = {
  restrictDataEdit: true,
  restrictStructureEdit: true,
  restrictScriptExecution: true,
  restrictDataImport: true,
};

const CONNECTION_READ_ONLY_TYPES = new Set([
  "mysql",
  "goldendb",
  "mariadb",
  "oceanbase",
  "diros",
  "starrocks",
  "sphinx",
  "postgres",
  "kingbase",
  "highgo",
  "vastbase",
  "opengauss",
  "gaussdb",
  "sqlserver",
  "iris",
  "sqlite",
  "duckdb",
  "oracle",
  "dameng",
  "tdengine",
  "clickhouse",
  "trino",
  "mongodb",
]);

const SQL_READ_ONLY_KEYWORDS = new Set([
  "select",
  "with",
  "show",
  "describe",
  "desc",
  "explain",
  "pragma",
  "values",
  "consume",
]);

const SQL_MUTATING_WITH_KEYWORDS = /\b(insert|update|delete|replace|merge|upsert)\b/i;
const SQL_SELECT_INTO_PATTERN = /^\s*select\b[\s\S]*\binto\b/i;

const MONGO_READ_ONLY_COMMANDS = new Set([
  "aggregate",
  "buildinfo",
  "collstats",
  "connectionstatus",
  "count",
  "countdocuments",
  "dbstats",
  "distinct",
  "explain",
  "find",
  "findone",
  "getparameter",
  "hello",
  "hostinfo",
  "ismaster",
  "listcollections",
  "listdatabases",
  "listindexes",
  "ping",
  "serverstatus",
]);

const MONGO_WRITE_COMMANDS = new Set([
  "bulkwrite",
  "collmod",
  "create",
  "createindexes",
  "delete",
  "drop",
  "dropdatabase",
  "dropindexes",
  "findandmodify",
  "insert",
  "mapreduce",
  "renamecollection",
  "update",
]);

const MONGO_META_KEYS = new Set([
  "$db",
  "$readpreference",
  "api",
  "apideprecationerrors",
  "apistrict",
  "comment",
  "let",
  "lsid",
  "maxtimems",
  "ordered",
  "readconcern",
  "writeconcern",
]);

const MYSQL_COMMENT_DIALECTS = new Set([
  "mysql",
  "goldendb",
  "mariadb",
  "oceanbase",
  "diros",
  "starrocks",
  "sphinx",
  "tidb",
]);

const isDashLineCommentStart = (text: string, dbType: string): boolean => {
  if (!MYSQL_COMMENT_DIALECTS.has(dbType)) return true;
  const next = text[2] || "";
  return !next || /\s/.test(next);
};

const supportsHashLineComment = (dbType: string): boolean =>
  !dbType || dbType === "clickhouse" || MYSQL_COMMENT_DIALECTS.has(dbType);

const isExecutableBlockComment = (text: string, dbType: string): boolean => {
  if (text.slice(0, 4).toLowerCase() === "/*m!") return dbType === "mariadb";
  return text.startsWith("/*!") && MYSQL_COMMENT_DIALECTS.has(dbType);
};

const stripLeadingSqlComments = (statement: string, dbType = ""): string => {
  let text = String(statement || "").trim();
  while (text) {
    if (text.startsWith("--") && isDashLineCommentStart(text, dbType)) {
      const next = text.indexOf("\n");
      text = next >= 0 ? text.slice(next + 1).trimStart() : "";
      continue;
    }
    if (text.startsWith("#") && supportsHashLineComment(dbType)) {
      const next = text.indexOf("\n");
      text = next >= 0 ? text.slice(next + 1).trimStart() : "";
      continue;
    }
    if (text.startsWith("/*") && !isExecutableBlockComment(text, dbType)) {
      const next = text.indexOf("*/");
      text = next >= 0 ? text.slice(next + 2).trimStart() : "";
      continue;
    }
    break;
  }
  return text;
};

const extractLeadingSqlKeyword = (statement: string, dbType = ""): string => {
  const text = stripLeadingSqlComments(statement, dbType);
  const match = text.match(/^[A-Za-z_][A-Za-z0-9_]*/);
  return match ? match[0].toLowerCase() : "";
};

const resolveConnectionReadOnlyType = (
  config: ConnectionReadOnlyLike,
): string => {
  if (!config) return "";
  return String(
    resolveSqlDialect(String(config.type || ""), String(config.driver || ""), {
      oceanBaseProtocol: config.oceanBaseProtocol,
    }),
  )
    .trim()
    .toLowerCase();
};

const isReadOnlySqlStatement = (statement: string, dbType: string): boolean => {
  const text = stripLeadingSqlComments(statement, dbType);
  if (!text) return true;
  const keyword = extractLeadingSqlKeyword(text, dbType);
  if (!keyword || !SQL_READ_ONLY_KEYWORDS.has(keyword)) {
    return false;
  }
  if (keyword === "select") {
    return !SQL_SELECT_INTO_PATTERN.test(text);
  }
  if (keyword === "with") {
    return !SQL_SELECT_INTO_PATTERN.test(text) &&
      !SQL_MUTATING_WITH_KEYWORDS.test(text);
  }
  return true;
};

const normalizeMongoCommandText = (statement: string): string => {
  const trimmed = String(statement || "").trim();
  if (!trimmed) return "";
  if (trimmed.startsWith("{")) {
    return trimmed;
  }
  const converted = convertMongoShellToJsonCommand(trimmed);
  if (converted.recognized && converted.command) {
    return converted.command;
  }
  return "";
};

const resolveMongoCommandKey = (command: Record<string, unknown>): string => {
  const keys = Object.keys(command).map((key) => key.trim());
  const effectiveKeys = keys.filter((key) => {
    const normalized = key.toLowerCase();
    return normalized !== "" && !MONGO_META_KEYS.has(normalized);
  });
  for (const key of effectiveKeys) {
    if (MONGO_WRITE_COMMANDS.has(key.toLowerCase())) {
      return key;
    }
  }
  for (const key of effectiveKeys) {
    if (MONGO_READ_ONLY_COMMANDS.has(key.toLowerCase())) {
      return key;
    }
  }
  return effectiveKeys[0] || "";
};

const isReadOnlyMongoStatement = (statement: string): boolean => {
  const commandText = normalizeMongoCommandText(statement);
  if (!commandText) return false;
  try {
    const parsed = JSON.parse(commandText) as Record<string, unknown>;
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      return false;
    }
    const commandKey = resolveMongoCommandKey(parsed).toLowerCase();
    if (!commandKey) return false;
    if (MONGO_WRITE_COMMANDS.has(commandKey)) {
      return false;
    }
    return MONGO_READ_ONLY_COMMANDS.has(commandKey);
  } catch {
    return false;
  }
};

const isConnectionReadOnlyStatement = (
  config: ConnectionReadOnlyLike,
  statement: string,
): boolean => {
  const dialect = resolveConnectionReadOnlyType(config);
  if (dialect === "mongodb") {
    return isReadOnlyMongoStatement(statement);
  }
  return isReadOnlySqlStatement(statement, dialect);
};

export const supportsConnectionReadOnlyMode = (
  config: ConnectionReadOnlyLike,
): boolean => {
  return CONNECTION_READ_ONLY_TYPES.has(resolveConnectionReadOnlyType(config));
};

export const normalizeConnectionProtectionConfig = (
  value: unknown,
): ConnectionProtectionConfig => {
  const raw =
    value && typeof value === "object"
      ? (value as Record<string, unknown>)
      : {};
  return {
    restrictDataEdit: raw.restrictDataEdit === true,
    restrictStructureEdit: raw.restrictStructureEdit === true,
    restrictScriptExecution: raw.restrictScriptExecution === true,
    restrictDataImport: raw.restrictDataImport === true,
  };
};

export const createEmptyConnectionProtectionConfig =
  (): ConnectionProtectionConfig => ({ ...EMPTY_CONNECTION_PROTECTION });

export const deriveLegacyConnectionReadOnlyFlag = (
  protection: unknown,
): boolean => {
  const normalized = normalizeConnectionProtectionConfig(protection);
  return CONNECTION_PROTECTION_KEYS.every((key) => normalized[key] === true);
};

export const resolveConnectionProtectionConfig = (
  config: ConnectionReadOnlyLike,
): ConnectionProtectionConfig => {
  if (!supportsConnectionReadOnlyMode(config)) {
    return createEmptyConnectionProtectionConfig();
  }
  const normalized = normalizeConnectionProtectionConfig(config?.protection);
  const hasExplicitRestriction = CONNECTION_PROTECTION_KEYS.some(
    (key) => normalized[key] === true,
  );
  if (hasExplicitRestriction) {
    return normalized;
  }
  if (config?.readOnly === true) {
    return { ...FULL_CONNECTION_PROTECTION };
  }
  return createEmptyConnectionProtectionConfig();
};

export const isConnectionProtectionEnabled = (
  config: ConnectionReadOnlyLike,
  key: ConnectionProtectionKey,
): boolean => {
  return resolveConnectionProtectionConfig(config)[key] === true;
};

export const getConnectionProtectionEnabledCount = (
  config: ConnectionReadOnlyLike,
): number => {
  const protection = resolveConnectionProtectionConfig(config);
  return CONNECTION_PROTECTION_KEYS.filter(
    (key) => protection[key] === true,
  ).length;
};

export const hasAnyConnectionProtection = (
  config: ConnectionReadOnlyLike,
): boolean => {
  return getConnectionProtectionEnabledCount(config) > 0;
};

export const isConnectionDataEditRestricted = (
  config: ConnectionReadOnlyLike,
): boolean => {
  return isConnectionProtectionEnabled(config, "restrictDataEdit");
};

export const isConnectionStructureEditRestricted = (
  config: ConnectionReadOnlyLike,
): boolean => {
  return isConnectionProtectionEnabled(config, "restrictStructureEdit");
};

export const isConnectionScriptExecutionRestricted = (
  config: ConnectionReadOnlyLike,
): boolean => {
  return isConnectionProtectionEnabled(config, "restrictScriptExecution");
};

export const isConnectionDataImportRestricted = (
  config: ConnectionReadOnlyLike,
): boolean => {
  return isConnectionProtectionEnabled(config, "restrictDataImport");
};

export const isConnectionForcedReadOnly = (
  config: ConnectionReadOnlyLike,
): boolean => {
  return deriveLegacyConnectionReadOnlyFlag(
    resolveConnectionProtectionConfig(config),
  );
};

export const findConnectionMutatingStatements = (
  config: ConnectionReadOnlyLike,
  sql: string,
): string[] => {
  if (!isConnectionScriptExecutionRestricted(config)) {
    return [];
  }
  return findSqlStatementRanges(String(sql || ""), resolveConnectionReadOnlyType(config))
    .map((range) => range.text.trim())
    .filter((statement) => statement.length > 0)
    .filter((statement) => !isConnectionReadOnlyStatement(config, statement));
};

File: internal/app/connection_readonly.go
package app

import (
	"encoding/json"
	"errors"
	"strings"

	"GoNavi-Wails/internal/connection"
)

type connectionProtectionKey string

const (
	connectionProtectionDataEdit        connectionProtectionKey = "restrictDataEdit"
	connectionProtectionStructureEdit   connectionProtectionKey = "restrictStructureEdit"
	connectionProtectionScriptExecution connectionProtectionKey = "restrictScriptExecution"
	connectionProtectionDataImport      connectionProtectionKey = "restrictDataImport"
)

var connectionReadOnlySupportedTypes = map[string]struct{}{
	"clickhouse": {},
	"dameng":     {},
	"diros":      {},
	"duckdb":     {},
	"gaussdb":    {},
	"highgo":     {},
	"iris":       {},
	"kingbase":   {},
	"mariadb":    {},
	"mongodb":    {},
	"mysql":      {},
	"oceanbase":  {},
	"opengauss":  {},
	"oracle":     {},
	"postgres":   {},
	"sphinx":     {},
	"sqlite":     {},
	"sqlserver":  {},
	"starrocks":  {},
	"tdengine":   {},
	"trino":      {},
	"vastbase":   {},
}

var mongoReadOnlyCommands = map[string]struct{}{
	"aggregate":        {},
	"buildinfo":        {},
	"collstats":        {},
	"connectionstatus": {},
	"count":            {},
	"countdocuments":   {},
	"dbstats":          {},
	"distinct":         {},
	"explain":          {},
	"find":             {},
	"findone":          {},
	"getparameter":     {},
	"hello":            {},
	"hostinfo":         {},
	"ismaster":         {},
	"listcollections":  {},
	"listdatabases":    {},
	"listindexes":      {},
	"ping":             {},
	"serverstatus":     {},
}

var mongoWriteCommands = map[string]struct{}{
	"bulkwrite":        {},
	"collmod":          {},
	"create":           {},
	"createindexes":    {},
	"delete":           {},
	"drop":             {},
	"dropdatabase":     {},
	"dropindexes":      {},
	"findandmodify":    {},
	"insert":           {},
	"mapreduce":        {},
	"renamecollection": {},
	"update":           {},
}

var mongoMetaCommandKeys = map[string]struct{}{
	"$db":                  {},
	"$readpreference":      {},
	"api":                  {},
	"apideprecationerrors": {},
	"apistrict":            {},
	"comment":              {},
	"let":                  {},
	"lsid":                 {},
	"maxtimems":            {},
	"ordered":              {},
	"readconcern":          {},
	"writeconcern":         {},
}

var readOnlyConnectionActionTextKeys = map[string]string{
	"创建数据库": "connection.backend.action.create_database",
	"connection.backend.action.create_database": "connection.backend.action.create_database",
	"创建模式": "connection.backend.action.create_schema",
	"connection.backend.action.create_schema":              "connection.backend.action.create_schema",
	"重命名模式":                                                "connection.backend.action.rename_schema",
	"connection.backend.action.rename_schema":              "connection.backend.action.rename_schema",
	"删除模式":                                                 "connection.backend.action.drop_schema",
	"connection.backend.action.drop_schema":                "connection.backend.action.drop_schema",
	"重命名数据库":                                               "connection.backend.action.rename_database",
	"connection.backend.action.rename_database":            "connection.backend.action.rename_database",
	"删除数据库":                                                "connection.backend.action.drop_database",
	"connection.backend.action.drop_database":              "connection.backend.action.drop_database",
	"重命名表":                                                 "connection.backend.action.rename_table",
	"connection.backend.action.rename_table":               "connection.backend.action.rename_table",
	"删除表":                                                  "connection.backend.action.drop_table",
	"connection.backend.action.drop_table":                 "connection.backend.action.drop_table",
	"删除视图":                                                 "connection.backend.action.drop_view",
	"connection.backend.action.drop_view":                  "connection.backend.action.drop_view",
	"删除函数或存储过程":                                            "connection.backend.action.drop_function_or_procedure",
	"connection.backend.action.drop_function_or_procedure": "connection.backend.action.drop_function_or_procedure",
	"重命名视图":                                                "connection.backend.action.rename_view",
	"connection.backend.action.rename_view":                "connection.backend.action.rename_view",
	"导入数据":                                                 "connection.backend.action.import_data",
	"connection.backend.action.import_data":                "connection.backend.action.import_data",
	"提交结果修改":                                               "connection.backend.action.apply_result_changes",
	"connection.backend.action.apply_result_changes":       "connection.backend.action.apply_result_changes",
	"预览结果修改":                                               "connection.backend.action.preview_result_changes",
	"connection.backend.action.preview_result_changes":     "connection.backend.action.preview_result_changes",
	"clear_table":                                          "connection.backend.action.clear_table",
	"connection.backend.action.clear_table":                "connection.backend.action.clear_table",
	"truncate_table":                                       "connection.backend.action.truncate_table",
	"connection.backend.action.truncate_table":             "connection.backend.action.truncate_table",
	"connection.backend.action.data_sync_structure":        "connection.backend.action.data_sync_structure",
	"数据同步写入":                                               "connection.backend.action.data_sync_write",
	"connection.backend.action.data_sync_write":            "connection.backend.action.data_sync_write",
}

func supportsConnectionReadOnlyMode(config connection.ConnectionConfig) bool {
	_, ok := connectionReadOnlySupportedTypes[resolveDDLDBType(config)]
	return ok
}

func hasAnyConnectionProtection(config connection.ConnectionProtectionConfig) bool {
	return config.RestrictDataEdit ||
		config.RestrictStructureEdit ||
		config.RestrictScriptExecution ||
		config.RestrictDataImport
}

func resolveConnectionProtectionConfig(config connection.ConnectionConfig) connection.ConnectionProtectionConfig {
	if !supportsConnectionReadOnlyMode(config) {
		return connection.ConnectionProtectionConfig{}
	}
	if hasAnyConnectionProtection(config.Protection) {
		return config.Protection
	}
	if config.ReadOnly {
		return connection.ConnectionProtectionConfig{
			RestrictDataEdit:        true,
			RestrictStructureEdit:   true,
			RestrictScriptExecution: true,
			RestrictDataImport:      true,
		}
	}
	return connection.ConnectionProtectionConfig{}
}

func isConnectionProtectionEnabled(config connection.ConnectionConfig, key connectionProtectionKey) bool {
	protection := resolveConnectionProtectionConfig(config)
	switch key {
	case connectionProtectionDataEdit:
		return protection.RestrictDataEdit
	case connectionProtectionStructureEdit:
		return protection.RestrictStructureEdit
	case connectionProtectionScriptExecution:
		return protection.RestrictScriptExecution
	case connectionProtectionDataImport:
		return protection.RestrictDataImport
	default:
		return false
	}
}

func isConnectionForcedReadOnly(config connection.ConnectionConfig) bool {
	protection := resolveConnectionProtectionConfig(config)
	return protection.RestrictDataEdit &&
		protection.RestrictStructureEdit &&
		protection.RestrictScriptExecution &&
		protection.RestrictDataImport
}

func isConnectionScriptExecutionRestricted(config connection.ConnectionConfig) bool {
	return isConnectionProtectionEnabled(config, connectionProtectionScriptExecution)
}

func normalizeReadOnlyConnectionText(text func(string, map[string]any) string) func(string, map[string]any) string {
	if text != nil {
		return text
	}
	return defaultAppText
}

func readOnlyConnectionQueryBlockedMessageWithText(text func(string, map[string]any) string) string {
	text = normalizeReadOnlyConnectionText(text)
	return text("query_editor.message.connection_readonly_blocked", nil)
}

func resolveReadOnlyConnectionActionLabel(action string, text func(string, map[string]any) string) string {
	label := strings.TrimSpace(action)
	if label == "" {
		return ""
	}
	text = normalizeReadOnlyConnectionText(text)
	if key, ok := readOnlyConnectionActionTextKeys[label]; ok {
		return text(key, nil)
	}
	return label
}

func readOnlyConnectionQueryBlockedMessage() string {
	return readOnlyConnectionQueryBlockedMessageWithText(defaultAppText)
}

func readOnlyConnectionActionBlockedMessageWithText(action string, text func(string, map[string]any) string) string {
	label := resolveReadOnlyConnectionActionLabel(action, text)
	if label == "" {
		return readOnlyConnectionQueryBlockedMessageWithText(text)
	}
	text = normalizeReadOnlyConnectionText(text)
	return text("connection.backend.error.readonly_action_blocked", map[string]any{"action": label})
}

func readOnlyConnectionActionBlockedMessage(action string) string {
	return readOnlyConnectionActionBlockedMessageWithText(action, defaultAppText)
}

func ensureConnectionAllowsQueryWithText(config connection.ConnectionConfig, query string, text func(string, map[string]any) string) error {
	text = normalizeReadOnlyConnectionText(text)
	if !isConnectionScriptExecutionRestricted(config) {
		return nil
	}
	for _, statement := range splitSQLStatements(query) {
		if trimmed := strings.TrimSpace(statement); trimmed != "" && !isReadOnlySQLQuery(resolveDDLDBType(config), trimmed) {
			return errors.New(readOnlyConnectionQueryBlockedMessageWithText(text))
		}
	}
	return nil
}

func ensureConnectionAllowsQuery(config connection.ConnectionConfig, query string) error {
	return ensureConnectionAllowsQueryWithText(config, query, defaultAppText)
}

func ensureConnectionAllowsActionWithText(config connection.ConnectionConfig, key connectionProtectionKey, action string, text func(string, map[string]any) string) error {
	text = normalizeReadOnlyConnectionText(text)
	if !isConnectionProtectionEnabled(config, key) {
		return nil
	}
	return errors.New(readOnlyConnectionActionBlockedMessageWithText(action, text))
}

func ensureConnectionAllowsAction(config connection.ConnectionConfig, key connectionProtectionKey, action string) error {
	return ensureConnectionAllowsActionWithText(config, key, action, defaultAppText)
}

func ensureConnectionAllowsDataEdit(config connection.ConnectionConfig, action string) error {
	return ensureConnectionAllowsAction(config, connectionProtectionDataEdit, action)
}

func ensureConnectionAllowsStructureEdit(config connection.ConnectionConfig, action string) error {
	return ensureConnectionAllowsAction(config, connectionProtectionStructureEdit, action)
}

func ensureConnectionAllowsDataImport(config connection.ConnectionConfig, action string) error {
	return ensureConnectionAllowsAction(config, connectionProtectionDataImport, action)
}

func isReadOnlyMongoCommand(query string) bool {
	trimmed := strings.TrimSpace(query)
	if !strings.HasPrefix(trimmed, "{") {
		return false
	}
	var doc map[string]interface{}
	if err := json.Unmarshal([]byte(trimmed), &doc); err != nil {
		return false
	}
	commandKey := resolveMongoCommandKey(doc)
	if commandKey == "" {
		return false
	}
	if _, blocked := mongoWriteCommands[commandKey]; blocked {
		return false
	}
	_, allowed := mongoReadOnlyCommands[commandKey]
	return allowed
}

func resolveMongoCommandKey(doc map[string]interface{}) string {
	commandKey := ""
	for key := range doc {
		normalized := strings.ToLower(strings.TrimSpace(key))
		if normalized == "" {
			continue
		}
		if _, isMeta := mongoMetaCommandKeys[normalized]; isMeta {
			continue
		}
		if _, isWrite := mongoWriteCommands[normalized]; isWrite {
			return normalized
		}
		if _, isRead := mongoReadOnlyCommands[normalized]; isRead {
			commandKey = normalized
		}
	}
	return commandKey
}
